### DISCUSSION TITLE
Question regarding removing percepts and triggering new events for subsequent percepts

### DISCUSSION THREAD

**Discussion by mindcrime** [jason-lang/jason#108] [Q&A] [answered]
Hi all, another n00bie question here if you will entertain it:

I have a simple Environment class that starts a Thread and emits an identical percept every 6 seconds (approximately). And some AgentSpeak code that just triggers when that percept appears as a belief and prints the time. I get that the "old" percept/belief has to be removed before a new event will be triggered, but there's a subtle detail (or two) that I am a little unclear about.

To wit:  If, in my Environment class I remove the percept using removePercept, AND in my AgentSpeak code I remove the belief using "-belief_name", then everything works exactly as I might expect. I see the belief show up and trigger an event approximately every 6 seconds.

**Environment**
```
    			while( true )
    			{
    				try
    				{
    					Thread.sleep(  6000 );
    				}
    				catch( Exception e )
    				{
    					e.printStackTrace();
    				}
    				
    				System.out.println( "adding percept \"oracle_killed\"" );
    				addPercept( "agent_smith", ok );
    				
    				
    				try
    				{
    					Thread.sleep(  500 );
    				}
    				catch( Exception e )
    				{
    					e.printStackTrace();
    				}
    				
    				System.out.println( "removing percept \"oracle_killed\"" );
    				removePercept( "agent_smith", ok );
    				
    			}
    			
    		}
    		
    	} );
```

***AgentSpeak***
```
started.

+started <- .print( "Hello, Mr. Anderson.").

+oracle_killed <- .time( HH, MM, SS ); .print( "The Oracle is dead, at: ", HH, ":", MM, ":", SS); .wait(500); -oracle_killed.
```

***Output***
```
^Cprhodes@troubadour:~/development/experimental/ai_exp/AgentSandbox/jason/simple_env_percept$ jason simple_env_percept.mas2j 
[agent_smith] Hello, Mr. Anderson.
[agent_smith] The Oracle is dead, at: 19:10:10
[agent_smith] The Oracle is dead, at: 19:10:16
[agent_smith] The Oracle is dead, at: 19:10:23
[agent_smith] The Oracle is dead, at: 19:10:29
[agent_smith] The Oracle is dead, at: 19:10:36

```

So that's all good. But if I don't do the -oracle_killed bit in the AgentSpeak code, I see some weird behavior. I still get the new events, but the timing becomes very chaotic.

***Output***
```
prhodes@troubadour:~/development/experimental/ai_exp/AgentSandbox/jason/simple_env_percept$ jason simple_env_percept.mas2j 
[agent_smith] Hello, Mr. Anderson.
[agent_smith] The Oracle is dead, at: 19:12:33
[agent_smith] The Oracle is dead, at: 19:12:59
[agent_smith] The Oracle is dead, at: 19:13:6
[agent_smith] The Oracle is dead, at: 19:13:12
[agent_smith] The Oracle is dead, at: 19:13:32
[agent_smith] The Oracle is dead, at: 19:13:38
```

and then in the other case I tried, I removed the removePercept() bit from the Environment and ONLY had the -oracle_killed part in the AgentSpeak and I never saw any event fired after the first.

So what is the takeaway from all this? Is it simply that when working with repeated percepts you should always both remove the belief explicitly in AgentSpeak code AND call removePercept() in the Environment code? Or is there some valid case for the scenario where you call removePercept() but don't remove the belief explicitly in the AgentSpeak code? If so, any details on how one should expect that work worK? 

And finally, I assume that it's the case that it's never sufficient to just do the -belief_name in the AgentSpeak code, in order to see the subsequent events fire? Is that correct, or am I missing something else.

Thanks for your patience as I work my way through all of this!

**Comment by jomifred** [comment, marked as answer]
Hi, some initial comments on that

- the `-oracle_killed` will *not* remove the belief! This command is interpreted as `-oracle_killed[source(self)]` while the belief is `oracle_killed[source(percept)]`. Either use `-oracle_killed[source(percept)]` or `.abolish(oracle_killed)`. Anyway, it does not seem to be the cause of issues.
- the Jason model for perception is that in the beginning of the reasoning cycle the agent takes a snapshot of current *state* of the environment. If the environment changes faster than the agent, it could be that the agent does not perceive some changes.  Other agent-environment interaction models have been proposed for Jason. For instance, in the Jason-CArtAgO case, the agent is not perceiving states, but a flow of *events* that changed the environment. In this case, no updates are lost. This model is available in JaCaMo platform. 

I can try to run your code later to see if I can identify the problem.

**Reply by jomifred** [reply]
I run your example. The reason for the strange behaviour is related to my second bullet above. The environment is changing while the agent is *sleeping*. Jason agents go for a nap if they have nothing to do  :-). To wake them up, you can call `informAgsEnvironmentChanged()` in your environment code.

**Reply by mindcrime** [reply]
`the -oracle_killed will not remove the belief! This command is interpreted as -oracle_killed[source(self)] while the belief is oracle_killed[source(percept)]. Either use -oracle_killed[source(percept)] or .abolish(oracle_killed). Anyway, it does not seem to be the cause of issues.`

Aaaah. Of course. I forgot about the source distinction. D'oh! 

`Jason agents go for a nap if they have nothing to do :-). To wake them up, you can call informAgsEnvironmentChanged() in your environment code.`

Gotcha. Thanks again!

`This model is available in JaCaMo platform.`

It's funny... I picked up the JaCaMo book a few weeks ago, not knowing much about JaCaMo at the time. When I realized that JaCaMo is largely based on Jason I decided to finally commit to truly learning Jason. Which led us all here. :-) So learning JaCaMo is definitely on "the list" but I'll work my way to it over time.
